### PR TITLE
api: StartFastCatchup initialize parameter

### PR DESCRIPTION
**Pull request by winder**
## Summary

Adds a new `initialize` query parameter to the `StartCatchup` admin API endpoint. It introduces a new concept for when a node is considered initialized compared to a catchpoint, which is defined by the value of the initialize parameter. The catchup request will be a no-op unless the catchpoint advances the node by at least `initialize` rounds.

## Test Plan

New unit test.

### CODE REVIEW

**Comment by jannotti** [comment]
Your reasoning for this made sense for adding this at the time, but now I can't recall it.  Why would the catchpoint endpoint be called if the node is already close to catchup?  I thought I remembered something like, "Once you're in catchup mode, you would keep trying to fast catchup." But that doesn't fit with what this does.

Is the problem that there are automated situations in which the catchpoint endpoint is called? Maybe in `sandbox`?

**Comment by winder** [comment]
> Your reasoning for this made sense for adding this at the time, but now I can't recall it. Why would the catchpoint endpoint be called if the node is already close to catchup? I thought I remembered something like, "Once you're in catchup mode, you would keep trying to fast catchup." But that doesn't fit with what this does.

It is mainly to simplify scripts/tools. i.e. when Conduit starts up, it has some logic to check the latest catchpoint and the latest round and it has to make a decision about whether or not to run fast-catchup. The docker container has a similar feature, but is a bash script. Since I'm now implementing this logic a 2nd time I figured it makes sense to put it in a common place.

With this feature, tools can run fast catchup unconditionally and things will just work. So it's harder to make a mistake, and they don't have to figure out our catchpoint label encoding.

**Comment by jannotti** [comment]
> > Your reasoning for this made sense for adding this at the time, but now I can't recall it. Why would the catchpoint endpoint be called if the node is already close to catchup? I thought I remembered something like, "Once you're in catchup mode, you would keep trying to fast catchup." But that doesn't fit with what this does.
> 
> @jannotti It is mainly to simplify scripts/tools. i.e. when Conduit starts up, it has some logic to check the latest catchpoint and the latest round and it has to make a decision about whether or not to run fast-catchup. The docker container has a similar feature, but is a bash script. Since I'm now implementing this logic a 2nd time I figured it makes sense to put it in a common place.
> 
> With this feature, tools can run fast catchup unconditionally and things will just work. So it's harder to make a mistake, and they don't have to figure out our catchpoint label encoding.

I see, that makes sense.  Is it worthwhile to make the boolean parameter an integer, instead of defining 1M as the unconditional distance?

**Comment by winder** [comment]
> > > Your reasoning for this made sense for adding this at the time, but now I can't recall it. Why would the catchpoint endpoint be called if the node is already close to catchup? I thought I remembered something like, "Once you're in catchup mode, you would keep trying to fast catchup." But that doesn't fit with what this does.
> > 
> > 
> > @jannotti It is mainly to simplify scripts/tools. i.e. when Conduit starts up, it has some logic to check the latest catchpoint and the latest round and it has to make a decision about whether or not to run fast-catchup. The docker container has a similar feature, but is a bash script. Since I'm now implementing this logic a 2nd time I figured it makes sense to put it in a common place.
> > With this feature, tools can run fast catchup unconditionally and things will just work. So it's harder to make a mistake, and they don't have to figure out our catchpoint label encoding.
> 
> I see, that makes sense. Is it worthwhile to make the boolean parameter an integer, instead of defining 1M as the unconditional distance?

I like that idea. I'll change this to an int.

**Comment by winder** [comment]
@jannotti your suggestion has been implemented, please take another look.

File: daemon/algod/api/server/v2/test/handlers_test.go [hunk 1]
```diff
@@ -1320,9 +1324,30 @@ func startCatchupTest(t *testing.T, catchpoint string, nodeError error, expected
 	req := httptest.NewRequest(http.MethodPost, "/", nil)
 	rec := httptest.NewRecorder()
 	c := e.NewContext(req, rec)
-	err := handler.StartCatchup(c, catchpoint)
+	var err error
+	if initRounds != 0 {
+		err = handler.StartCatchup(c, catchpoint, model.StartCatchupParams{Initialize: &initRounds})
+	} else {
+		err = handler.StartCatchup(c, catchpoint, model.StartCatchupParams{})
+	}
```
Comment (jannotti) [inline daemon/algod/api/server/v2/test/handlers_test.go:1332]: With the parameter change, this all becomes.
```suggestion
	err := handler.StartCatchup(c, catchpoint, model.StartCatchupParams{Initialize: &initRounds})
```

File: daemon/algod/api/server/v2/handlers.go [hunk 2]
```diff
@@ -1393,12 +1393,22 @@ func (v2 *Handlers) getPendingTransactions(ctx echo.Context, max *uint64, format
...
 // startCatchup Given a catchpoint, it starts catching up to this catchpoint
-func (v2 *Handlers) startCatchup(ctx echo.Context, catchpoint string) error {
-	_, _, err := ledgercore.ParseCatchpointLabel(catchpoint)
+func (v2 *Handlers) startCatchup(ctx echo.Context, catchpoint string, initializeRounds uint64) error {
+	catchpointRound, _, err := ledgercore.ParseCatchpointLabel(catchpoint)
 	if err != nil {
 		return badRequest(ctx, err, errFailedToParseCatchpoint, v2.Log)
 	}
 
+	if initializeRounds > 0 {
```
[the remark this answers is not in the document]
Comment (winder) [reply daemon/algod/api/server/v2/handlers.go:1402]: You're allowed to use fast-catchup to move the node backwards, so we still need this.

File: daemon/algod/api/server/v2/test/handlers_test.go (hunk 1 above)
Comment (winder) [reply daemon/algod/api/server/v2/test/handlers_test.go:1332]: I left this in so that the `nil` parameter handling would still be exercised

**Review by jannotti — APPROVED** [review]

**Comment by jannotti** [comment]
Will autogen SDKs automatically pick up the new parameter?

**Comment by winder** [comment]
> Will autogen SDKs automatically pick up the new parameter?

No, the SDKs don't support admin endpoints like this one.

File: daemon/algod/api/server/v2/handlers.go [hunk 3]
```diff
@@ -1393,12 +1393,22 @@ func (v2 *Handlers) getPendingTransactions(ctx echo.Context, max *uint64, format
...
+		ledgerRound := v2.Node.LedgerForAPI().Latest()
```
Comment (algorandskiy) [inline daemon/algod/api/server/v2/handlers.go:1403]: I think need a test for this condition to have this behavior "fixed".

Comment (winder) [reply daemon/algod/api/server/v2/handlers.go:1403]: There's a new unit test in this PR. How would you like it to be improved?

Comment (algorandskiy) [reply daemon/algod/api/server/v2/handlers.go:1403]: I mismatched `"the node has already been initialized"` check and the  error`errCatchpointWouldNotInitialize`. Looks all right.

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**